File: stonedb/clog.go
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

package stonedb

import (
	"sync/atomic"
)

func (db *DB) buildSnapshotLocked() Snapshot {
	xip := make(map[uint64]bool, len(db.activeXids))
	for xid := range db.activeXids {
		xip[xid] = true
	}
	xmax := atomic.LoadUint64(&db.transactionID) + 1
	return Snapshot{Xmax: xmax, Xip: xip}
}

func (db *DB) clogStatus(xid uint64) TxStatus {
	db.txMu.Lock()
	_, inProgress := db.activeXids[xid]
	db.txMu.Unlock()
	if inProgress {
		return TxInProgress
	}
	db.clogMu.RLock()
	st, ok := db.clog[xid]
	db.clogMu.RUnlock()
	if !ok {
		return TxCommitted
	}
	return st
}

func (db *DB) setClog(xid uint64, status TxStatus) {
	db.clogMu.Lock()
	db.clog[xid] = status
	db.clogMu.Unlock()
}

func (db *DB) forgetClog(xid uint64) {
	db.clogMu.Lock()
	delete(db.clog, xid)
	db.clogMu.Unlock()
}

func (db *DB) isVisible(xmin uint64, snap Snapshot) bool {
	if xmin >= snap.Xmax || snap.contains(xmin) {
		return false
	}
	return db.clogStatus(xmin) == TxCommitted
}

func (db *DB) abortTransaction(tx *Transaction) {
	if tx.xid == 0 {
		return
	}

	tx.abortOnce.Do(func() {
		if !tx.claimDecision() {
			return
		}

		db.txMu.Lock()
		_, stillActive := db.activeXids[tx.xid]
		db.txMu.Unlock()
		if !stillActive {
			return
		}

		if _, err := db.appendRecord(WALRecordAbort, tx.xid, nil, nil); err != nil {
			panic("CRITICAL: ABORT record append failed: " + err.Error())
		}
		db.setClog(tx.xid, TxAborted)
		db.index.DropXid(tx.xid)

		db.txMu.Lock()
		delete(db.activeXids, tx.xid)
		delete(db.beginOffsets, tx.xid)
		delete(db.txStartTimes, tx.xid)
		for k := range tx.keyLocks {
			if o, ok := db.keyLocks[k]; ok && o == tx.xid {
				delete(db.keyLocks, k)
			}
		}
		db.txMu.Unlock()
		db.forgetClog(tx.xid)
	})
}

File: rust/crates/turnstone-engine/src/clog.rs
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

use std::collections::{HashMap, HashSet};
use std::sync::RwLock;

use crate::types::{Snapshot, TxStatus};

/// Commit-log and active-transaction tracking (subset of Go `clog.go`).
#[derive(Default)]
pub struct ClogState {
    pub active_xids: RwLock<HashSet<u64>>,
    clog: RwLock<HashMap<u64, TxStatus>>,
}

impl ClogState {
    pub fn build_snapshot(&self, transaction_id: u64) -> Snapshot {
        let xmax = transaction_id + 1;
        let active = self.active_xids.read().unwrap();
        if active.is_empty() {
            return Snapshot {
                xmax,
                xip: HashMap::new(),
            };
        }
        let xip = active.iter().map(|&xid| (xid, true)).collect();
        Snapshot { xmax, xip }
    }

    pub fn clog_status(&self, xid: u64) -> TxStatus {
        if self.active_xids.read().unwrap().contains(&xid) {
            return TxStatus::InProgress;
        }
        if let Some(st) = self.clog.read().unwrap().get(&xid) {
            return *st;
        }
        TxStatus::Committed
    }

    pub fn set_clog(&self, xid: u64, status: TxStatus) {
        self.clog.write().unwrap().insert(xid, status);
    }

    pub fn forget_clog(&self, xid: u64) {
        self.clog.write().unwrap().remove(&xid);
    }

    pub fn clog_len(&self) -> usize {
        self.clog.read().unwrap().len()
    }

    pub fn is_visible(&self, xmin: u64, snap: &Snapshot) -> bool {
        if xmin >= snap.xmax || snap.contains(xmin) {
            return false;
        }
        self.clog_status(xmin) == TxStatus::Committed
    }
}
